File: javascript/14-messenger/test/notification_client.test.js
import assert from 'node:assert';
import test from 'node:test';
import { NotificationClient } from '../src/notification_client.js';

test('sends notification through layers', () => {
  const client = new NotificationClient();
  const result = client.send({ recipient: 'user-1', message: 'Hello', channel: 'sms' });
  assert.strictEqual(result.status, 'sent');
  assert.strictEqual(result.delivery_id, 'sms-user-1');
  assert.strictEqual(result.payload.message, 'Hello');
});

test('defaults channel and priority', () => {
  const client = new NotificationClient();
  const result = client.send({ message: 'Test' });
  assert.strictEqual(result.status, 'sent');
  assert.strictEqual(result.payload.channel, 'email');
  assert.strictEqual(result.payload.priority, 'normal');
});

test('preserves explicit priority', () => {
  const client = new NotificationClient();
  const result = client.send({ recipient: 'ops', message: 'Urgent', priority: 'high' });
  assert.strictEqual(result.payload.priority, 'high');
  assert.strictEqual(result.payload.recipient, 'ops');
});

test('rejects missing message', () => {
  const client = new NotificationClient();
  const result = client.send({ recipient: 'ops' });
  assert.strictEqual(result.status, 'rejected');
  assert.strictEqual(result.reason, 'missing_message');
});

test('records observable audit events', () => {
  const client = new NotificationClient();
  const result = client.send({ message: 'Deploy complete', channel: 'push' });
  assert.deepStrictEqual(result.audit, ['queued:push', 'sent:push']);
});

test('reports failed delivery for unsupported channel', () => {
  const client = new NotificationClient();
  const result = client.send({ recipient: 'ops', message: 'Legacy alert', channel: 'fax' });
  assert.strictEqual(result.status, 'failed');
  assert.strictEqual(result.reason, 'unsupported_channel');
  assert.deepStrictEqual(result.audit, ['queued:fax', 'failed:fax']);
});

test('high priority failed delivery is scheduled for retry', () => {
  const client = new NotificationClient();
  const result = client.send({ recipient: 'ops', message: 'Legacy alert', channel: 'fax', priority: 'high' });
  assert.strictEqual(result.status, 'retrying');
  assert.strictEqual(result.reason, 'unsupported_channel');
  assert.deepStrictEqual(result.audit, ['queued:fax', 'failed:fax', 'retry_scheduled:fax']);
});

File: csharp/14-messenger/test/notification_client_test.cs
using Xunit;

public class NotificationClientTest
{
    [Fact]
    public void SendsNotificationThroughLayers()
    {
        var client = new NotificationClient();
        var payload = new Dictionary<string, object> { { "recipient", "user-1" }, { "message", "Hello" }, { "channel", "sms" } };
        var result = client.send(payload);
        Assert.Equal("sent", result.Status);
        Assert.Equal("sms-user-1", result.DeliveryId);
        Assert.Equal("Hello", result.Payload["message"]);
    }

    [Fact]
    public void DefaultsChannelAndPriority()
    {
        var client = new NotificationClient();
        var result = client.send(new Dictionary<string, object> { { "message", "Test" } });
        Assert.Equal("sent", result.Status);
        Assert.Equal("email", result.Payload["channel"]);
        Assert.Equal("normal", result.Payload["priority"]);
    }

    [Fact]
    public void PreservesExplicitPriority()
    {
        var client = new NotificationClient();
        var payload = new Dictionary<string, object> { { "recipient", "ops" }, { "message", "Urgent" }, { "priority", "high" } };
        var result = client.send(payload);
        Assert.Equal("high", result.Payload["priority"]);
        Assert.Equal("ops", result.Payload["recipient"]);
    }

    [Fact]
    public void RejectsMissingMessage()
    {
        var client = new NotificationClient();
        var result = client.send(new Dictionary<string, object> { { "recipient", "ops" } });
        Assert.Equal("rejected", result.Status);
        Assert.Equal("missing_message", result.Reason);
    }

    [Fact]
    public void RecordsObservableAuditEvents()
    {
        var client = new NotificationClient();
        var result = client.send(new Dictionary<string, object> { { "message", "Deploy complete" }, { "channel", "push" } });
        Assert.Equal(new List<string> { "queued:push", "sent:push" }, result.Audit);
    }

    [Fact]
    public void ReportsFailedDeliveryForUnsupportedChannel()
    {
        var client = new NotificationClient();
        var result = client.send(new Dictionary<string, object> { { "recipient", "ops" }, { "message", "Legacy alert" }, { "channel", "fax" } });
        Assert.Equal("failed", result.Status);
        Assert.Equal("unsupported_channel", result.Reason);
        Assert.Equal(new List<string> { "queued:fax", "failed:fax" }, result.Audit);
    }

    [Fact]
    public void HighPriorityFailedDeliveryIsScheduledForRetry()
    {
        var client = new NotificationClient();
        var result = client.send(new Dictionary<string, object> { { "recipient", "ops" }, { "message", "Legacy alert" }, { "channel", "fax" }, { "priority", "high" } });
        Assert.Equal("retrying", result.Status);
        Assert.Equal("unsupported_channel", result.Reason);
        Assert.Equal(new List<string> { "queued:fax", "failed:fax", "retry_scheduled:fax" }, result.Audit);
    }
}
